File: Copy_List_with_Random_Pointer.cpp
#include <iostream>
#include <map>
using namespace std;
// * Definition for singly-linked list with a random pointer.
struct RandomListNode {
    int label;
    RandomListNode *next, *random;
    RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
};

class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
        if(head==nullptr) return nullptr;
        map<RandomListNode*,RandomListNode*> old2new;
        RandomListNode *newhead=new RandomListNode(head->label);
        old2new[head]=newhead;
        RandomListNode *old_=head, *new_=newhead;
        while(old_ != nullptr){
            if(old_->next==nullptr)
                new_->next = nullptr;
            else if(old2new.find(old_->next)!=old2new.end())
                new_->next = old2new[old_->next];
            else{
                new_->next = new RandomListNode(old_->next->label);
                old2new[old_->next]=new_->next;
            }

            if(old_->random==nullptr)
                new_->random = nullptr;
            else if(old2new.find(old_->random)!=old2new.end())
                new_->random = old2new[old_->random];
            else{
                new_->random = new RandomListNode(old_->random->label);
                old2new[old_->random]=new_->random;
            }
            
            old_ = old_->next;
            new_ = new_->next;
        }

        return newhead;
    }
};

int main() {
    #define N new RandomListNode
    Solution s;
    s.copyRandomList(N(1));
    return 0;
}

File: Algorithms/Copy_List_with_Random_Pointer.py
# Definition for singly-linked list with a random pointer.
class RandomListNode:
    def __init__(self, x):
        self.label = x
        self.next = None
        self.random = None

class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):
        if head is None:
            return None
        old2new = {}
        newhead = RandomListNode(head.label)
        old2new[head]=newhead
        old,new = head,newhead
        while old is not None:
            if old.next is None:
                new.next = None
            elif old.next in old2new:
                new.next = old2new[old.next]
            else:
                new.next = RandomListNode(old.next.label)
                old2new[old.next]=new.next

            if old.random is None:
                new.random = None
            elif old.random in old2new:
                new.random = old2new[old.random]
            else:
                new.random = RandomListNode(old.random.label)
                old2new[old.random]=new.random
            old = old.next
            new = new.next
        return newhead

if __name__=="__main__":
    s=Solution()
    N = RandomListNode
    l = s.copyRandomList(N(1))
